align_system/utils/incontext_utils.py

```diff
@@ -0,0 +1,287 @@
+            
+            # TODO - add KDMA normalization option
+            
+            return incontext_data
+
+    def select_icl_examples(self, sys_kdma_name, prompt_to_match):
+        '''
+        Selects a list of relevant ICL examples
+        Input:
+            sys_kdma_name - key of the target kdma in self.icl_datasets
+            prompt_to_match - the prompt we are selecting ICL examples for
+        Output:
+            selected_icl_examples - relevant subset of self.icl_datasets
+        '''
+        # Check that we have incontext examples for the target kdma
+        if sys_kdma_name not in self.icl_datasets:
+            raise RuntimeError(f"No incontext samples for targeted kdma: {sys_kdma_name}")
+        n_icl_examples = self.incontext_settings["number"]
+        possible_icl_examples = self.icl_datasets[sys_kdma_name]
+        # Check that we have enough incontext examples for the target kdma
+        if len(possible_icl_examples) < n_icl_examples:
+            raise RuntimeError(f"Not enough possible incontext samples to learn from. Only "
+                            f"{len(possible_icl_examples)} samples available while asking for "
+                            f"{n_icl_examples} incontext samples.")
+        # If using LOO, don't include example ICL with exact same prompt
+        if self.incontext_settings.get("leave_one_out", False):
+            possible_icl_examples = [
+                icl_ex for icl_ex in possible_icl_examples
+                if icl_ex["prompt"] != prompt_to_match
+            ]
+
+        # Downselect to n_icl_examples via given method
+        icl_strategy = self.incontext_settings["method"]
+        
+        if icl_strategy == "random":
+            selected_icl_examples = random.sample(possible_icl_examples, n_icl_examples)
+        elif icl_strategy == "bert_similarity":
+            possible_icl_prompts = [icl_sample["prompt"] for icl_sample in possible_icl_examples]
+            # Create similarity scores between the ICL samples and find top-k indices
+            from bert_score import score
+            _, _, F1 = score([prompt_to_match]*len(possible_icl_prompts), possible_icl_prompts, lang="en")
+            _, indices = torch.topk(F1, n_icl_examples)
+
+            selected_icl_examples = [possible_icl_examples[i] for i in indices]
+        else:
+            raise ValueError(f'"{icl_strategy}" is not a valid incontext method. Please use "random" or '
+                                '"bert_similarity"')
+
+        return selected_icl_examples
+
+
+class ComparativeRegressionIncontextExampleGenerator(IncontextExampleGenerator):
+    def __init__(self,
+                 incontext_settings, 
+                 target_kdmas, 
+                 log,
+                 **kwargs):
+        self.incontext_settings = incontext_settings
+        self.target_kdmas = target_kdmas
+        self.log = log
+        self.set_icl_datasets()
```

Comment:
I'm of the opinion that we shouldn't redefine this since it's doing the exact same thing as the base class, but open to hearing other opinions. At a minimum, I'm thinking it should be changed to `super().__init__()`?

Reply:
I didn't mean to copy `__init__` here, I agree we don't need it. Removed here https://github.com/ITM-Kitware/align-system/pull/115/commits/0ee831d9e81ad1dc08103214be8d0019dfe56da1